### DISCUSSION TITLE
Are id fields in fieldArray items removed after altering the array?

### DISCUSSION THREAD

**Discussion by stomvi** [react-hook-form/react-hook-form#6307] [Q&A]
CSB:
[https://codesandbox.io/s/react-hook-form-nested-field-array-forked-d10sf?file=/src/ParentFieldArray.js](https://codesandbox.io/s/react-hook-form-nested-field-array-forked-d10sf?file=/src/ParentFieldArray.js)

While `watch` the `options` field, `id` fields get removed after the array is altered using `append` or `remove`.  Is this in purpose? If other fields change depend on the array and the `id` values are crucial, do I have to duplicate to some other keys to keep the values?

**Comment by bluebill1049** [comment]
`id` is made for `key` not for you to store the value, if you need `id`, you probably should keep a copy of your own.

**Reply by stomvi** [reply]
Thanks for your quick reply. I think it's worthy to mention this in the document.

The "mark as answer" button is down. I will try later.

**Comment by stomvi** [comment]
I found out hook form fills the `id` with UUID while it is absent in the last version and publish it along with other fields, but not anymore?

From the document, I know `id` is required and provide my own.  And other fields are depending on those `id`s.  I never though that they won't show up in the watched value.

**Comment by stomvi** [comment]
I would like to continue this as a discussion.

I feel like `id` field is quite generic and is too common to use as internal uses.  In cases that need to update array records according to its ids, sending modified data along with the `id` is intuitive.  According to you answer, If users have to transform backend data and copy `id` value to another field, they have to transform it back before it is sent to the backend for many reasons, including lack of rights to alter backend API.  And we are using the same `Joi` schema both at the frontend and backend.  Adding up those conditions, the problem left is, when to transform the data.

Still it would be very nice to remain the `id` fields in the result value sine more and more projects keep their `id` unique throughout the systems.  I genuinely hope this would change in the next update.

Appreciations for your great job.

**Reply by bluebill1049** [reply]
you can change the `keyName` to others.

**Reply by stomvi** [reply]
WOW what a rapid reply.  Thank you!
